pennylane/transforms/qfunc_transforms.py

```diff
@@ -0,0 +1,386 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains tools and decorators for registering qfunc transforms."""
+# pylint: disable=too-few-public-methods
+import functools
+import inspect
+
+import pennylane as qml
+
+
+def make_tape(fn):
+    """Returns a function that generates the tape from a quantum function without any
+    operation queuing taking place.
+
+    This is useful when you would like to manipulate or transform
+    the tape created by a quantum function without evaluating it.
+
+    Args:
+        fn (function): the quantum function to generate the tape from
+
+    Returns:
+        function: The returned function takes the same arguments as the quantum
+        function. When called, it returns the generated quantum tape
+        without any queueing occuring.
+
+    **Example**
+
+    Consider the following quantum function:
+
+    .. code-block:: python
+
+        def qfunc(x):
+            qml.Hadamard(wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.RX(x, wires=0)
+
+    We can use ``make_tape`` to extract the tape generated by this
+    quantum function, without any of the operations being queued by
+    any existing queuing contexts:
+
+    >>> with qml.tape.QuantumTape() as active_tape:
+    ...     qml.RY(1.0, wires=0)
+    ...     tape = make_tape(qfunc)(0.5)
+    >>> tape.operations
+    [Hadamard(wires=[0]), CNOT(wires=[0, 1]), RX(0.5, wires=[0])]
+
+    Note that the currently recording tape did not queue any of these quantum operations:
+
+    >>> active_tape.operations
+    [RY(1.0, wires=[0])]
+    """
+
+    def wrapper(*args, **kwargs):
+        active_tape = qml.tape.get_active_tape()
+
+        if active_tape is not None:
+            with active_tape.stop_recording(), active_tape.__class__() as tape:
+                fn(*args, **kwargs)
+        else:
+            with qml.tape.QuantumTape() as tape:
+                fn(*args, **kwargs)
+        return tape
+
+    return wrapper
+
+
+class NonQueuingTape(qml.queuing.AnnotatedQueue):
+    """Mixin class that creates a tape that does not queue
+    itself to the current queuing context."""
+
+    def _process_queue(self):
+        super()._process_queue()
+
+        for obj, info in self._queue.items():
+            qml.queuing.QueuingContext.append(obj, **info)
+
+        qml.queuing.QueuingContext.remove(self)
+
+
+class single_tape_transform:

```

Comment:
I did not notice this before, but why is this a class instead of a function like `qfunc_transform`?

Reply:
purely so that when we need to, we can do `isinstance(fn, single_tape_transform)` 😆 So it's an implementation detail